Update event loop documentation for gui_qt deprecation (#2639)
Napari/napari: docs/guides/event_loop.md

# An Introduction to the Event Loop in napari

## Brief summary

It is not necessary to have a deep understanding of Qt or event loops to use
napari. napari attempts to use "sane defaults" for most scenarios. Here are the
most important details:

### In IPython or Jupyter Notebook

> napari will detect if you are running an an IPython or Jupyter shell, and will

Reviewer: ```suggestion
napari will detect if you are running an IPython or Jupyter shell, and will
```

automatically use the [IPython GUI event
loop](https://ipython.readthedocs.io/en/stable/config/eventloops.html#integrating-with-gui-event-loops).

> As of [version 0.4.7](https://github.com/napari/napari/releases/tag/v0.4.7) is

Reviewer: ```suggestion
As of [version 0.4.7](https://github.com/napari/napari/releases/tag/v0.4.7), it is
```

In [1]: import napari

In [2]: viewer = napari.Viewer()  # Viewer will show in a new window

In [3]: ... # Continue interactive usage
```

````{tip}
If you would prefer that napari did *not* start the interactive
event loop for you in IPython, then you can disable it with:

```python
from napari.utils import SETTINGS

SETTINGS.application.ipy_interactive = False
```

... but then you will have to start the program yourself as described [below](#in-a-script).
````

### In a script

Outside of IPython, you must tell napari when to "start the program" using
{func}`napari.run`.  This will *block* execution of your script at that point,
show the viewer, and wait for any user interaction.  When the last viewer
closes, execution of the script will proceed.

```python
import napari

viewer = napari.Viewer()
...  # Continue setting  up your program

# Start the program, show the viewer, wait for GUI interaction.
napari.run() 

# Anything below here will execute only after the viewer is closed.
```

-----------

## More in depth...

Like most applications with a graphical user interface (GUI), napari operates
within an **event loop** that waits for – and responds to – events triggered by

> the user interacting with the program.  These events might be something like a

Reviewer: ```suggestion
the user's interactions with the interface.  These events might be something like a
```

# An Introduction to the Event Loop in napari

> ## tldr;

Reviewer 2: Can we have a less casual heading here? Maybe "Brief summary" or something else.

> ## tldr;

Author: ```suggestion
## Brief summary
```

It is not necessary to have a deep understanding of Qt or event loops to use

> napari.  napari attempts to use "sane defaults" for most scenarios. Here are the

Reviewer 3: ```suggestion
napari. napari attempts to use "sane defaults" for most scenarios. Here are the
```

most important details:

### In IPython or Jupyter Notebook

napari will detect if you are running an an IPython or Jupyter shell, and will
automatically use the [IPython GUI event
loop](https://ipython.readthedocs.io/en/stable/config/eventloops.html#integrating-with-gui-event-loops).

> As of [version 0.4.7](https://github.com/napari/napari/releases/tag/v0.4.7) is

Reviewer 2: I'd prefer the docs focus on what we do now, and de-emphasize what we used to do.

It's important to give people the info that we no longer need to call `%gui qt`, but this is probably better distributed in the release notes, on zulip, and twitter, etc.

Author: I did try to deemphasize it with the offset, but given the 2 year usage... I kinda thought a lot of people would be asking "what about gui_qt??"  So it'd be nice to send them somewhere to answer that

Author: (Also not needing `%gui qt` magic is a different point than this one, which is about `napari.gui_qt`)

Reviewer 4: I think this is ok for a transition period, but maybe in 3-6 months we can revisit and update the language a bit

Author: yeah, definitely doesn't have to be there forever

no longer necessary to call `%gui qt` manually.  Just create a viewer:

```python
In [1]: import napari

In [2]: viewer = napari.Viewer()  # viewer will show in a new window

> In [3]: ... # continue interactive usage

Reviewer 3: I suggest we capitalize comments?

It kinda feels like a cleaner convention (that we do not enforce really... but we could :) )

so 

```python

# Continue interactive usage

vs

# continue interactive usage

```

```

````{tip}
If you would prefer that napari did *not* start the interactive
event loop for you in IPython, then you can disable it with:

```python
from napari.utils import SETTINGS

SETTINGS.application.ipy_interactive = False
```

But then you will have to start the program yourself as described [below](#in-a-script).

````
### In a script

Outside of IPython, you must tell napari when to "start the program" using
{func}`napari.run`.  This will *block* execution of your script at that point,
show the viewer, and wait for any user interaction.  When the last viewer
closes, execution of the script will proceed.

```python
import napari

viewer = napari.Viewer()

> ... # continue setting  up your program

Reviewer 3: Same here...

> # start the program, show the viewer, wait for GUI interaction.

Reviewer 3: Same here...

napari.run() 

> # anything below here will execute only after the viewer is closed.

Reviewer 3: Same here...

```

-----------

> 

Reviewer 2: Can we shift the section on  "Hooking up your own events" (which includes sub-sections on keybindings, mouse events, et) up above the nitty gritty about "The Qt Event Loop".

I actually think it might also be good to move the in-depth Qt Event Loop stuff into a separate markdown file, and just link to it from here for anyone who wants more information. That would make it much clearer that these parts are "explainer" documentation on how things work under the hood. Mixing it in here with the other practical info, like how to make your own keybindings, is very likely to confuse people.

Author: I kinda feel like if anything should go in another file, it's the hooking up of custom events (that always felt out of place to me here).  The details about the qt event loop are the direct explanation of _why_ you need napari.run sometimes and not other times (which I think you had asked for elsewhere?)

Reviewer 2: Yeah, these things should get split up.

Reviewer 2: I think that's a good idea, and it's also a good idea to clearly distinguish (a) which bits of the docs people need to use/type, vs (b) the bits we are explaining so you understand the stuff under the hood better (but we don't generally want new users to try and sprinkle `QApplication` stuff through their code)

napari.run()  # This will call `app.exec_()` and start the event loop.
```

(gui-qt-deprecated)=

:::{admonition}  What about `napari.gui_qt`?
:class: caution

**{func}`napari.gui_qt` was deprecated in version 0.4.8.**

The autocreation of the `QApplication` instance and the {func}`napari.run`
function was introduced in
[PR#2056](https://github.com/napari/napari/pull/2056), and released in [version
0.4.3](https://github.com/napari/napari/releases/tag/v0.4.3).  Prior to that,
all napari examples included this `gui_qt()` context manager:

```python
# deprecated
with napari.gui_qt():
    viewer = napari.Viewer()
```

On entering the context, `gui_qt` would create a `QApplication`, and on exiting
the context, it would start the event loop (the two critical steps [mentioned
above](#qt-applications-and-event-loops)).  

Unlike a typical context manager, however, it did not actually *destroy* the
`QApplication` (since it may still be needed in the same session)... and future
calls to `gui_qt` were only needed to start the event loop.  By auto-creating
the `QApplication` during {class}`~napari.Viewer` creation, introducing the

> explicit {func}`napari.run`, and using the [integrated IPython event

Reviewer: ```suggestion
explicit {func}`napari.run` function, and using the [integrated IPython event
```

            process_event(event)
```

Actions taken by the user add events to the queue (e.g. "button pressed",
"slider moved", etc...), and the event loop handles them one at a time.

## Qt Applications and Event Loops

Currently, napari uses Qt as its GUI backend, and the main loop handling events
in napari is the [Qt
EventLoop](https://wiki.qt.io/Threads_Events_QObjects#Events_and_the_event_loop).

A deep dive into the Qt event loop is beyond the scope of this document, but
it's worth being aware of two critical steps in the "lifetime" of a Qt
Application:

1) Any program that would like to create a
   [`QWidget`](https://doc.qt.io/qt-5/qwidget.html) (the class from which all
   napari's graphical elements are subclassed), must create a
   [`QApplication`](https://doc.qt.io/qt-5/qapplication.html) instance *before*
   instantiating any widgets.

   ```python

>    from qtpy.QtWidgets import QApplication()

Reviewer 3: ```suggestion
   from qtpy.QtWidgets import QApplication
```

>    app = QApplication([])

Reviewer 3: Maybe add a comment that `[]` are the args passed to the app or something?

Just, to not leave that hanging in there without an explanation :)

   ```

2) In order to actually show and interact with widgets, one must start the
   application's event loop:

   ```python
   app.exec_()
   ```

### napari's `QApplication`

In napari, the initial step of creating the `QApplication` is handled by
{func}`napari.qt.get_app`.  (Note however, that napari will do this for you
automatically behind the scenes when you create a viewer with
{class}`napari.Viewer()`)

The second step – starting the Qt event loop – is handled by {func}`napari.run`

```python
import napari

> viewer = napari.Viewer()  # this will create a Application if one doesn't exist

Reviewer 3: Capitalize comments... maybe... if you agree 🤷🏼

> napari.run()  # this will call `app.exec_()` and start the event loop.

Reviewer 3: same here 🙃
